# Notebook 02 — Limpieza y validación inicial de datos

## Índice

1. Análisis datos tabla: `customers`  
2. Análisis datos tabla: `geolocations`  
3. Análisis datos tabla: `order_items`  
4. Análisis datos tabla: `order_payments`  
5. Análisis datos tabla: `order_reviews`  
6. Análisis datos tabla: `orders`  
7. Análisis datos tabla: `products`  
8. Análisis datos tabla: `sellers`  
9. Síntesis general de transformaciones realizadas  
10. Preparación para la integración de datos


## Objetivo del notebook

El objetivo de este notebook es realizar una primera limpieza y validación de calidad de los datasets del marketplace. Para cada tabla se revisan su estructura, dimensiones, tipos de datos, valores nulos, duplicados y posibles incoherencias básicas.

La limpieza se plantea de forma conservadora: no se eliminan registros salvo que sean claramente redundantes o inconsistentes. En los casos en los que los valores nulos representan ausencia de información válida o eventos no ocurridos, se conservan y se documenta su significado. El objetivo principal es preparar los datos para su posterior integración y análisis logístico, evitando introducir supuestos artificiales en esta fase inicial.

In [1]:
import pandas as pd
import os
import numpy as np

In [2]:
os.chdir("..")
os.getcwd()

'c:\\Users\\lusoz\\Desktop\\tfg'

In [3]:
# ALL DATA
customers = pd.read_csv("data/raw/datasets-Brazilian E-Commerce Public Dataset by Olist/olist_customers_dataset.csv")
geolocations = pd.read_csv("data/raw/datasets-Brazilian E-Commerce Public Dataset by Olist/olist_geolocation_dataset.csv")
order_items = pd.read_csv("data/raw/datasets-Brazilian E-Commerce Public Dataset by Olist/olist_order_items_dataset.csv")
order_payments = pd.read_csv("data/raw/datasets-Brazilian E-Commerce Public Dataset by Olist/olist_order_payments_dataset.csv")
order_reviews = pd.read_csv("data/raw/datasets-Brazilian E-Commerce Public Dataset by Olist/olist_order_reviews_dataset.csv")
order_dataset = pd.read_csv("data/raw/datasets-Brazilian E-Commerce Public Dataset by Olist/olist_orders_dataset.csv")
products = pd.read_csv("data/raw/datasets-Brazilian E-Commerce Public Dataset by Olist/olist_products_dataset.csv")
sellers = pd.read_csv("data/raw/datasets-Brazilian E-Commerce Public Dataset by Olist/olist_sellers_dataset.csv")


# 1. Análisis datos tabla: customers

In [4]:
# visualización previa del dataframe.
customers.head()

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP


In [5]:
# Visualización de la forma del dataframe.
customers.shape

(99441, 5)

In [6]:
# Visualización de información del dataframe.
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB


In [7]:
#visualización de valores nulos en el dataframe.
customers.isnull().sum()

customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

In [8]:
# visualización de valores duplicados en el dataframe.
customers.duplicated().sum()

np.int64(0)

In [9]:
customers["customer_city"].value_counts()

customer_city
sao paulo                   15540
rio de janeiro               6882
belo horizonte               2773
brasilia                     2131
curitiba                     1521
                            ...  
olhos d'agua                    1
pacotuba                        1
sao sebastiao do paraiba        1
benedito leite                  1
campos verdes                   1
Name: count, Length: 4119, dtype: int64

In [10]:
# Visualización de espacios vacíos en el dataframe.
(customers == "").sum()

customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

In [11]:
# Limpieza de datos
# todos los valores de la columna "customer_city" a minúsculas y eliminar espacios en blanco al inicio y al final.
customers["customer_city"] = customers["customer_city"].str.lower().str.strip()
# todos los valores de la columna "customer_state" a mayúsculas y eliminar espacios en blanco al inicio y al final.
customers["customer_state"] = customers["customer_state"].str.upper().str.strip()
# convertir la columna "customer_zip_code_prefix" a tipo de dato string.
customers["customer_zip_code_prefix"] = customers["customer_zip_code_prefix"].astype(str)

#### Revisión inicial de la tabla `customers`

Se realizó una primera exploración de la tabla `customers` mediante `head()`, `shape`, `info()`, `isnull().sum()`, `duplicated().sum()`, `nunique()`, `value_counts()` para la variable `customer_city` y una comprobación adicional de cadenas vacías.

La tabla contiene 99.441 registros y 5 variables: `customer_id`, `customer_unique_id`, `customer_zip_code_prefix`, `customer_city` y `customer_state`.

No se detectaron valores nulos ni filas duplicadas. Además, la revisión de cadenas vacías no mostró campos aparentemente vacíos codificados como texto. Los identificadores y las variables geográficas presentan una estructura coherente para su uso posterior en el análisis.

En consecuencia, la tabla `customers` no requiere una limpieza intensiva. Las únicas transformaciones recomendables son la normalización preventiva de las variables de texto y el tratamiento de `customer_zip_code_prefix` como variable categórica, ya que representa un código territorial y no una magnitud numérica.

# 2. Revisión tabla: geolocations

In [12]:
# visualizacion del dataframe
geolocations.head()

,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,1037,-23.545621,-46.639292,sao paulo,SP
1,1046,-23.546081,-46.644820,sao paulo,SP
2,1046,-23.546129,-46.642951,sao paulo,SP
3,1041,-23.544392,-46.639499,sao paulo,SP
4,1035,-23.541578,-46.641607,sao paulo,SP


In [13]:
# Visualización de la forma del dataframe.
geolocations.shape

(1000163, 5)

In [14]:
# visualización de información del dataframe.
geolocations.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   geolocation_zip_code_prefix  1000163 non-null  int64  
 1   geolocation_lat              1000163 non-null  float64
 2   geolocation_lng              1000163 non-null  float64
 3   geolocation_city             1000163 non-null  object 
 4   geolocation_state            1000163 non-null  object 
dtypes: float64(2), int64(1), object(2)
memory usage: 38.2+ MB


In [15]:
# visualización de valores nulos en el dataframe.
geolocations.isnull().sum()

geolocation_zip_code_prefix    0
geolocation_lat                0
geolocation_lng                0
geolocation_city               0
geolocation_state              0
dtype: int64

In [16]:
# visualización de valores duplicados en el dataframe.
geolocations.duplicated().sum()

np.int64(261831)

In [17]:
# eliminar duplicados exactos
geolocations = geolocations.drop_duplicates().copy()
geolocations.duplicated().sum()

np.int64(0)

In [18]:
#  Eliminar duplicados exactos y crear nueva variable geolocations_clean
geolocations_clean = geolocations.drop_duplicates().copy()

#  Ver si siguen existiendo múltiples filas por código postal
geolocations_clean["geolocation_zip_code_prefix"].duplicated().sum()

np.int64(719317)

In [19]:
#  Agrupar por código postal y calcular la media de latitud y longitud, y tomar la primera ciudad y estado de los duplicados.
geolocations_zip = (
    geolocations_clean
    .groupby("geolocation_zip_code_prefix", as_index=False)
    .agg({
        "geolocation_lat": "mean",
        "geolocation_lng": "mean",
        "geolocation_city": "first",
        "geolocation_state": "first"
    })
)

geolocations_zip.shape


(19015, 5)

In [20]:
# Ver si siguen existiendo múltiples filas por código postal
geolocations_zip["geolocation_zip_code_prefix"].duplicated().sum()

np.int64(0)

#### Limpieza y tratamiento de la tabla `geolocations`

La tabla `geolocations` contiene información geográfica asociada a prefijos de código postal. A diferencia de otras tablas del dataset, esta tabla no está vinculada directamente a un identificador individual de cliente, sino que puede relacionarse con la tabla `customers` mediante la coincidencia entre `customer_zip_code_prefix` y `geolocation_zip_code_prefix`.

En la revisión inicial no se detectaron valores nulos, pero sí un número elevado de duplicados exactos. Estos duplicados correspondían a filas que repetían la misma combinación de prefijo postal, latitud, longitud, ciudad y estado, por lo que fueron considerados registros redundantes y eliminados sin pérdida de información analítica.

Tras eliminar los duplicados exactos, se comprobó que muchos prefijos de código postal seguían apareciendo varias veces con coordenadas distintas. Esto indica que un mismo prefijo postal puede estar asociado a varias observaciones geográficas dentro de una misma zona. Para evitar que una futura unión con la tabla de clientes multiplicase artificialmente el número de registros, se construyó una versión agregada de la tabla con una sola fila por prefijo postal.

En esta tabla agregada, cada `geolocation_zip_code_prefix` queda representado por la latitud y longitud medias de sus observaciones, conservando además una ciudad y un estado de referencia. El resultado fue una tabla con 19.015 prefijos postales únicos y sin duplicados en la variable utilizada para la unión.

Es importante señalar que estas coordenadas no representan la ubicación exacta del domicilio del cliente, sino una localización aproximada asociada a su zona postal. Por tanto, su uso es adecuado para análisis territoriales agregados, visualización geográfica o estimación aproximada de distancias logísticas, pero no para una geolocalización individual precisa.

 # 3. Revisión tabla: order_items

In [21]:
# visualización previa del dataframe.
order_items.head()

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14


In [22]:
# Visualización de la forma del dataframe.
order_items.shape

(112650, 7)

In [23]:
# visualización de información del dataframe.
order_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB


In [24]:
#visualización de valores nulos en el dataframe.
order_items.isnull().sum()

order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

In [25]:
#visualización de valores duplicados en el dataframe.
order_items.duplicated().sum()

np.int64(0)

In [26]:
# cambiar formato para la fecha de la columna "shipping_limit_date" a tipo de dato datetime.
order_items["shipping_limit_date"] = pd.to_datetime(order_items["shipping_limit_date"])

In [27]:
# comprobar nulos en la conlumna "shipping_limit_date" tras el cambio de formato.
order_items["shipping_limit_date"].isnull().sum()

np.int64(0)

In [28]:
# revisar valores negativos en la columna "price" y "freight_value"
(order_items["price"] < 0).sum(), (order_items["freight_value"] < 0).sum()

(np.int64(0), np.int64(0))

#### Limpieza y revisión de la tabla `order_items`

La tabla `order_items` contiene información a nivel de línea de pedido. Es decir, cada registro no representa necesariamente un pedido completo, sino un producto concreto dentro de un pedido. Por este motivo, un mismo `order_id` puede aparecer varias veces si el pedido contiene más de un producto.

En la revisión inicial se comprobó la estructura de la tabla mediante `head()`, `shape`, `info()`, `isnull().sum()` y `duplicated().sum()`. La tabla contiene 112.650 registros y 7 variables: `order_id`, `order_item_id`, `product_id`, `seller_id`, `shipping_limit_date`, `price` y `freight_value`.

No se detectaron valores nulos ni filas duplicadas exactas, por lo que no fue necesario eliminar registros por ausencia de información o repetición completa de datos.

Sin embargo, se observó que la variable `shipping_limit_date` estaba almacenada como tipo `object`, aunque representa una fecha. Por ello, se convirtió a formato `datetime`, permitiendo su uso posterior en análisis temporales relacionados con los plazos de envío. Tras la conversión, se comprobó que no se habían generado valores nulos, lo que indica que todas las fechas pudieron transformarse correctamente.

También se revisaron las variables económicas `price` y `freight_value`. La variable `price` representa el precio del producto, mientras que `freight_value` representa el coste de envío o transporte asociado al ítem. Se comprobó que no existían valores negativos en ninguna de estas dos columnas, lo que refuerza la coherencia básica de los datos económicos.

La variable `order_item_id`, aunque aparece como numérica, se interpreta como un identificador secuencial del ítem dentro de cada pedido, no como una magnitud cuantitativa. Por tanto, no se considera una variable sobre la que tenga sentido aplicar operaciones estadísticas directas, sino una clave auxiliar para distinguir productos dentro de un mismo pedido.

En conjunto, la tabla `order_items` presenta una estructura limpia y coherente. La principal transformación aplicada fue la conversión de `shipping_limit_date` a formato temporal, junto con la validación de ausencia de nulos, duplicados y valores económicos negativos.

# 4. Análisis datos tabla: order_payments

In [29]:
order_payments.head()

,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39
2,25e8ea4e93396b6fa0d3dd708e76c1bd,1,credit_card,1,65.71
3,ba78997921bbcdc1373bb41e913ab953,1,credit_card,8,107.78
4,42fdf880ba16b47b59251dd489d4441a,1,credit_card,2,128.45


In [30]:
order_payments.shape

(103886, 5)

In [31]:
order_payments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  object 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  object 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 4.0+ MB


In [32]:
order_payments.isnull().sum()

order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64

In [33]:
order_payments.duplicated().sum()

np.int64(0)

In [34]:
# revisar valores negativos y de pago 0 en payment_value
(order_payments["payment_value"] < 0).sum(), (order_payments["payment_value"] == 0).sum()

(np.int64(0), np.int64(9))

In [35]:
# revisar valores negativos en la columna "payment_installments"
(order_payments["payment_installments"] < 0).sum()

np.int64(0)

#### Limpieza y revisión de la tabla `order_payments`

La tabla `order_payments` contiene información sobre los pagos asociados a los pedidos. Cada registro representa una operación de pago vinculada a un `order_id`, incluyendo el tipo de pago, la secuencia del pago, el número de cuotas y el valor económico pagado.

En primer lugar, se realizó una exploración general de la tabla mediante `head()`, `shape` e `info()`. La tabla contiene 103.886 registros y 5 variables: `order_id`, `payment_sequential`, `payment_type`, `payment_installments` y `payment_value`.

La revisión de valores nulos mostró que no existen campos vacíos en ninguna de las columnas. Asimismo, la comprobación de duplicados exactos indicó que no hay filas completamente repetidas, por lo que no fue necesario eliminar registros por duplicidad.

Posteriormente, se revisaron las variables numéricas con posible impacto analítico. En el caso de `payment_value`, no se encontraron valores negativos, lo cual confirma una coherencia básica en los importes de pago. Sin embargo, se detectaron algunos registros con valor de pago igual a cero. Estos casos no se eliminaron en esta fase, ya que aparecen asociados principalmente a métodos como `voucher` o a pagos con tipo `not_defined`, por lo que pueden representar situaciones especiales del sistema de pago y no necesariamente errores de carga.

También se comprobó que la variable `payment_installments`, que representa el número de cuotas del pago, no contiene valores negativos. Esta variable se mantiene como numérica, ya que puede ser útil posteriormente para analizar patrones de financiación o comportamiento de pago.

La variable `payment_sequential` se interpreta como el orden o secuencia del pago dentro de un mismo pedido. Por tanto, aunque aparece como variable entera, no se considera una magnitud económica, sino un identificador ordinal que permite distinguir pagos múltiples asociados a un mismo `order_id`.

En conjunto, la tabla `order_payments` presenta una estructura adecuada para su integración posterior con la tabla principal de pedidos. No se realizaron transformaciones adicionales, ya que no se detectaron problemas graves de formato, ausencia de datos o incoherencias económicas relevantes.

# 5. Análisis datos tabla: order_review

In [36]:
order_reviews.head()

,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,NaN,NaN,2018-01-18 00:00:00,2018-01-18 21:46:59
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,NaN,NaN,2018-03-10 00:00:00,2018-03-11 03:05:13
2,228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,NaN,NaN,2018-02-17 00:00:00,2018-02-18 14:36:24
3,e64fb393e7b32834bb789ff8bb30750e,658677c97b385a9be170737859d3511b,5,NaN,Recebi bem antes do prazo estipulado.,2017-04-21 00:00:00,2017-04-21 22:02:06
4,f7c4243c7fe1938f181bec41a392bdeb,8e6bfb81e283fa7e4f11123a3fb894f1,5,NaN,Parabéns lojas lannister adorei comprar pela I...,2018-03-01 00:00:00,2018-03-02 10:26:53


In [37]:
order_reviews.shape

(99224, 7)

In [38]:
order_reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   review_id                99224 non-null  object
 1   order_id                 99224 non-null  object
 2   review_score             99224 non-null  int64 
 3   review_comment_title     11568 non-null  object
 4   review_comment_message   40977 non-null  object
 5   review_creation_date     99224 non-null  object
 6   review_answer_timestamp  99224 non-null  object
dtypes: int64(1), object(6)
memory usage: 5.3+ MB


In [39]:
order_reviews.isnull().sum()

review_id                      0
order_id                       0
review_score                   0
review_comment_title       87656
review_comment_message     58247
review_creation_date           0
review_answer_timestamp        0
dtype: int64

In [40]:
order_reviews.duplicated().sum()

np.int64(0)

In [41]:
# cambiar formato para la fecha de la columna "review_creation_date" y "review_answer_timestamp" a tipo de dato datetime.
order_reviews["review_creation_date"] = pd.to_datetime(
    order_reviews["review_creation_date"],
    errors="coerce"
)

order_reviews["review_answer_timestamp"] = pd.to_datetime(
    order_reviews["review_answer_timestamp"],
    errors="coerce"
)

order_reviews[["review_creation_date", "review_answer_timestamp"]].isnull().sum()

review_creation_date       0
review_answer_timestamp    0
dtype: int64

In [42]:
# comprobar valores únicos en la columna "review_score" no mayores de 5 y no menores de 1.
order_reviews["review_score"].value_counts().sort_index()

review_score
1    11424
2     3151
3     8179
4    19142
5    57328
Name: count, dtype: int64

#### Limpieza y revisión de la tabla `order_reviews`

La tabla `order_reviews` contiene información sobre las valoraciones realizadas por los clientes después de sus pedidos. Cada registro incluye el identificador de la reseña, el pedido asociado, la puntuación numérica, posibles comentarios textuales y las fechas de creación y respuesta de la valoración.

En primer lugar, se realizó una exploración general mediante `head()`, `shape`, `info()`, `isnull().sum()` y `duplicated().sum()`. La tabla contiene 99.224 registros y 7 variables: `review_id`, `order_id`, `review_score`, `review_comment_title`, `review_comment_message`, `review_creation_date` y `review_answer_timestamp`.

No se detectaron valores nulos en las variables principales de identificación, puntuación ni fechas. Tampoco se encontraron filas duplicadas exactas, por lo que no fue necesario eliminar registros por duplicidad.

Sí se observaron valores nulos en las columnas `review_comment_title` y `review_comment_message`. Estos nulos no se consideran errores, ya que muchos clientes pueden dejar únicamente una puntuación numérica sin escribir un título o comentario textual. Por este motivo, se decidió conservar estos registros, dado que siguen aportando información relevante a través de la variable `review_score`.

Las columnas `review_creation_date` y `review_answer_timestamp` estaban almacenadas como tipo `object`, aunque representan información temporal. Por ello, fueron convertidas a formato `datetime`. Tras la conversión, se comprobó que no se generaron valores nulos adicionales, lo que indica que todas las fechas pudieron transformarse correctamente.

También se revisó la distribución de `review_score`, comprobando que las puntuaciones se encuentran dentro del rango esperado de 1 a 5. Esta variable será útil posteriormente como indicador de satisfacción del cliente y podrá relacionarse con variables logísticas como tiempos de entrega, retrasos o incidencias.

No se aplicó una limpieza avanzada sobre los textos de las reseñas, ya que el objetivo principal del análisis no es realizar procesamiento de lenguaje natural. No obstante, se conserva la información textual disponible por si pudiera ser útil en fases posteriores del análisis.

# 6. Análisis datos tabla: Order_dataset

In [43]:
order_dataset.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00


In [44]:
order_dataset.shape

(99441, 8)

In [67]:
order_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   order_id                       99441 non-null  object        
 1   customer_id                    99441 non-null  object        
 2   order_status                   99441 non-null  object        
 3   order_purchase_timestamp       99441 non-null  datetime64[ns]
 4   order_approved_at              99281 non-null  datetime64[ns]
 5   order_delivered_carrier_date   97658 non-null  datetime64[ns]
 6   order_delivered_customer_date  96476 non-null  datetime64[ns]
 7   order_estimated_delivery_date  99441 non-null  object        
dtypes: datetime64[ns](4), object(4)
memory usage: 6.1+ MB


In [71]:
order_dataset.isnull().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

In [47]:
order_dataset.duplicated().sum()

np.int64(0)

In [70]:
order_dataset["order_purchase_timestamp"] = pd.to_datetime(order_dataset["order_purchase_timestamp"])
order_dataset["order_approved_at"] = pd.to_datetime(order_dataset["order_approved_at"])
order_dataset["order_delivered_carrier_date"] = pd.to_datetime(order_dataset["order_delivered_carrier_date"])
order_dataset["order_delivered_customer_date"] = pd.to_datetime(order_dataset["order_delivered_customer_date"])
order_dataset["order_estimated_delivery_date"] = pd.to_datetime(order_dataset["order_estimated_delivery_date"])

# 7. Análisis datos tabla: product

In [48]:
products.head()

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0
3,cef67bcfe19066a932b7673e239eb23d,bebes,27.0,261.0,1.0,371.0,26.0,4.0,26.0
4,9dc1a7de274444849c219cff195d0b71,utilidades_domesticas,37.0,402.0,4.0,625.0,20.0,17.0,13.0


In [49]:
products.shape

(32951, 9)

In [50]:
products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32341 non-null  object 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB


In [51]:
products.isnull().sum()

product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

In [52]:
products.duplicated().sum()

np.int64(0)

In [53]:
physical_cols = [
    "product_weight_g",
    "product_length_cm",
    "product_height_cm",
    "product_width_cm"
]

for col in physical_cols:
    print(col)
    print("Negativos:", (products[col] < 0).sum())
    print("Ceros:", (products[col] == 0).sum())

product_weight_g
Negativos: 0
Ceros: 4
product_length_cm
Negativos: 0
Ceros: 0
product_height_cm
Negativos: 0
Ceros: 0
product_width_cm
Negativos: 0
Ceros: 0


In [54]:
products.loc[products["product_weight_g"] == 0, "product_weight_g"] = np.nan

### Limpieza y revisión de la tabla `products`

La tabla `products` contiene información descriptiva y física de los productos del marketplace. En la revisión inicial se comprobó su estructura mediante `shape`, `info()`, `isnull().sum()` y `duplicated().sum()`. La tabla contiene 32.951 registros y 9 variables, sin filas duplicadas exactas.

Se detectaron valores nulos en dos grupos de variables. Por un lado, aparecen 610 valores nulos en variables catalográficas como `product_category_name`, `product_name_lenght`, `product_description_lenght` y `product_photos_qty`. Estos nulos parecen corresponder a productos con información descriptiva incompleta. No se eliminaron estos registros, ya que podrían estar asociados a pedidos reales y seguir siendo relevantes para el análisis logístico.

Por otro lado, se identificaron 2 valores nulos en las variables físicas del producto: `product_weight_g`, `product_length_cm`, `product_height_cm` y `product_width_cm`. Estas variables son especialmente relevantes desde el punto de vista logístico, ya que el peso y las dimensiones pueden influir en el coste de envío, la dificultad de transporte y los tiempos de entrega.

Además de los valores nulos, se revisó la existencia de valores negativos o iguales a cero en las variables físicas. No se encontraron valores negativos en ninguna de ellas, lo que indica una coherencia básica en las medidas registradas. Sin embargo, se detectaron 4 productos con `product_weight_g` igual a 0. Dado que un producto físico no debería tener un peso real de 0 gramos, estos casos se interpretaron como valores anómalos o como información faltante codificada incorrectamente.

Por este motivo, los valores de `product_weight_g` iguales a 0 fueron convertidos a `NaN`, tratándolos como valores ausentes. No se aplicaron correcciones adicionales sobre las dimensiones del producto, ya que `product_length_cm`, `product_height_cm` y `product_width_cm` no presentaron valores negativos ni iguales a cero.

En conjunto, la tabla `products` presenta una estructura adecuada para su integración posterior con la tabla `order_items`. La limpieza realizada fue conservadora, evitando eliminar productos y corrigiendo únicamente aquellos valores físicos claramente inconsistentes.

# 8. Análisis datos tabla: sellers 

In [55]:
sellers.head()

,seller_id,seller_zip_code_prefix,seller_city,seller_state
0,3442f8959a84dea7ee197c632cb2df15,13023,campinas,SP
1,d1b65fc7debc3361ea86b5f14c68d2e2,13844,mogi guacu,SP
2,ce3ad9de960102d0677a81f5d0bb7b2d,20031,rio de janeiro,RJ
3,c0f3eea2e14555b6faeea3dd58c1b1c3,4195,sao paulo,SP
4,51a04a8a6bdcb23deccc82b0b80742cf,12914,braganca paulista,SP


In [56]:
sellers.shape

(3095, 4)

In [57]:
sellers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3095 entries, 0 to 3094
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   seller_id               3095 non-null   object
 1   seller_zip_code_prefix  3095 non-null   int64 
 2   seller_city             3095 non-null   object
 3   seller_state            3095 non-null   object
dtypes: int64(1), object(3)
memory usage: 96.8+ KB


In [58]:
sellers.isnull().sum()

seller_id                 0
seller_zip_code_prefix    0
seller_city               0
seller_state              0
dtype: int64

In [59]:
sellers.duplicated().sum()

np.int64(0)

In [60]:
sellers["seller_id"].duplicated().sum()

np.int64(0)

In [61]:
sellers["seller_zip_code_prefix"] = sellers["seller_zip_code_prefix"].astype(str)

sellers["seller_city"] = sellers["seller_city"].str.lower().str.strip()
sellers["seller_state"] = sellers["seller_state"].str.upper().str.strip()

#### Limpieza y revisión de la tabla `sellers`

La tabla `sellers` contiene información básica sobre los vendedores del marketplace. Incluye el identificador del vendedor, su prefijo de código postal, ciudad y estado.

En la revisión inicial se comprobó la estructura de la tabla mediante `head()`, `shape`, `info()`, `isnull().sum()` y `duplicated().sum()`. La tabla contiene 3.095 registros y 4 variables: `seller_id`, `seller_zip_code_prefix`, `seller_city` y `seller_state`.

No se detectaron valores nulos ni filas duplicadas exactas, por lo que no fue necesario eliminar registros. La variable `seller_id` funciona como identificador del vendedor y permitirá unir esta tabla con `order_items`, donde también aparece el vendedor asociado a cada línea de pedido.

Aunque `seller_zip_code_prefix` aparece originalmente como variable numérica, se interpreta como una variable categórica/geográfica, ya que representa un prefijo postal y no una magnitud cuantitativa. Por este motivo, se recomienda convertirla a formato texto antes de futuras integraciones con tablas geográficas.

También se aplicó una normalización preventiva sobre las variables territoriales, eliminando espacios en blanco y homogeneizando el formato de ciudad y estado. En conjunto, la tabla `sellers` presenta una estructura limpia y adecuada para su integración posterior en el análisis logístico.

# Síntesis general de la limpieza realizada

A lo largo del notebook se realizó una revisión inicial de las principales tablas del dataset de e-commerce: `customers`, `geolocations`, `order_items`, `order_payments`, `order_reviews`, `order_dataset`, `products` y `sellers`.

En todas las tablas se aplicó un procedimiento común de exploración: revisión de las primeras filas, dimensiones del dataframe, tipos de datos, valores nulos, duplicados exactos y coherencia básica de las variables. Esta revisión permitió identificar qué tablas requerían transformaciones específicas y cuáles presentaban ya una estructura suficientemente limpia.

Las transformaciones aplicadas fueron principalmente de cuatro tipos. En primer lugar, se normalizaron variables textuales como ciudades y estados, eliminando espacios innecesarios y homogeneizando mayúsculas y minúsculas. En segundo lugar, se convirtieron variables temporales almacenadas como texto a formato `datetime`, como `shipping_limit_date`, `review_creation_date` y `review_answer_timestamp`, para permitir análisis temporales posteriores.

En tercer lugar, se trataron variables categóricas codificadas como numéricas, especialmente los prefijos de código postal, que fueron interpretados como identificadores geográficos y no como magnitudes cuantitativas. En cuarto lugar, se revisaron variables económicas y físicas para detectar valores negativos, valores iguales a cero o posibles incoherencias. En el caso de `products`, los pesos iguales a cero fueron convertidos a valores ausentes, al considerarse incompatibles con productos físicos.

La tabla `geolocations` requirió un tratamiento específico, ya que contenía un elevado número de duplicados exactos y múltiples coordenadas para un mismo prefijo postal. Para evitar problemas en futuras integraciones, se eliminaron duplicados exactos y se construyó una tabla agregada por código postal, usando la latitud y longitud medias como localización aproximada de cada zona postal.

En algunos casos se conservaron valores nulos por razones metodológicas. Por ejemplo, en `order_reviews`, los nulos en título o comentario de la reseña no se consideran errores, sino ausencia de texto por parte del cliente. Del mismo modo, en `products`, los nulos catalográficos se mantuvieron para no eliminar productos que podrían estar asociados a pedidos reales. En `order_dataset`, los nulos en fechas logísticas deben interpretarse con cautela, ya que pueden representar eventos no ocurridos o no registrados, como pedidos no aprobados, no enviados o no entregados.

En conjunto, la limpieza realizada prepara los datos para la siguiente fase del proyecto: la integración de tablas y la construcción de variables analíticas orientadas al rendimiento logístico, como tiempos de entrega, retrasos, costes de envío, características físicas del producto, ubicación aproximada de clientes y vendedores, y satisfacción del cliente.